Resolving bug that prevents ssl_verify option for Unifi device_tracker

GenericStudent opened the pull request:
## Description:
The unifi device_tracker takes a conf option `verify_ssl` which the docs state to accept: True, False, or an absolute path to a CA cert bundle for SSL verification. This is not the case as the code explicitly does verification on the conf option as a **boolean**.

I've updated the tests which should pass once I get the correct voluptuous config validators in place.

My question is how do I validate the CONF_VERIFY_SSL using vol and ensuring the given config value is either True, False, or a valid path? (I'd settle on string)

**Related issue (if applicable):** fixes #7892

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** N/A

## Example entry for `configuration.yaml` (if applicable):
```yaml
- platform: unifi
  interval_seconds: 30
  verify_ssl: "/home/homeassistant/.homeassistant/unifi-ca-chain.crt"
  host: unifi.<my domain>
  port: 8443
  username: !secret unifi_username
  password: !secret unifi_password
```

## Checklist:

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

GenericStudent committed:
Added TODO to illustrate my intentions

homeassistant commented:
Hi @GenericStudent, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#9788).

Once you do that we will be able to review and accept this pull request.

Thanks!

GenericStudent committed:
Resolved linting issue

MartinHjelmare commented on homeassistant/components/device_tracker/unifi.py:
```diff
@@ -26,13 +26,14 @@
     vol.Required(CONF_USERNAME): cv.string,
     vol.Required(CONF_PORT, default=DEFAULT_PORT): cv.port,
-    vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean,
+    vol.Optional(CONF_VERIFY_SSL, default=DEFAULT_VERIFY_SSL): cv.boolean
```
Use `vol.Any(cv.boolean, cv.string)`.

MartinHjelmare replied:
Perhaps use `vol.IsFile()` instead of `cv.string`.

GenericStudent replied:
Hi Martin, thanks for the heads up on this! I've implemented isfile and mocked the tests to pass for "valid" files existing and failing for non-existing files. My question now is I'm trying to have it fail tests with the verify_ssl set to incorrect types such as int, but this is passing through the boolean validator. Should I only test for True, False, paths; or should I be testing incorrect types? If so how do I get boolean to be more rigid / explicit?

MartinHjelmare replied:
Is it really necessary to have it more rigid? The validation should make sure that errors and configuration mistakes are avoided. It doesn't have to catch all cases, I think, where the impact is low. I'd test for the cases that should work and the ones you want to avoid.

GenericStudent replied:
Fair points. I'll remove the type checking tests and push up. Thanks again for your help.

sdague replied:
The new tests below seem to work for me if this is set to vol.Any(cv.boolean, cv.string). That's still better than the current code and will fix the feature.

GenericStudent replied:
Just pushed the fix that allows a file to be passed. I've removed the two tests you mentioned @sdague as they were failing to raise due to boolean accepting `None` and `5555` values in favour of file existence validation. Do you think I should change it to cv.string instead? Essentially which validation is greater value, file exists or valid bool | string?

sdague reviewed (dismissed):
Still need to fix this to allow strings.

GenericStudent committed:
Resolved bool or file validation and updated tests

The tests have been updated to include mocks to assert a temp
ca cert exists as it should for the positive tests with an
additional negative test for a file not existing being tested.

sdague commented:
It feels like bool | string is probably good enough. It would fix the issue you were trying to fix. I'm going to assume a bad file error will be pretty clear in the logs later.

GenericStudent commented:
So those tests were only passing due to a missing PASSWORD config, putting that back leads to it not raising config errors on the `None` and `5555` `CONF_VERIFY_SSL` directive, should I redact out the `isfile` changes and remove the extra tests?

GenericStudent committed:
Resolved flake8 linting issues (test docstrings)

GenericStudent commented:
@sdague I'm happy with the PR now unless you want the above changed?

pvizeli commented:
Good work :+1: `cv.isFile` is better as `cv.string`
